runner/src/server/plugins/clientSideUpload.ts

```diff
@@ -0,0 +1,73 @@
+import { HapiRequest, HapiResponseToolkit } from "../types";
+
+export default {
+  plugin: {
+    name: "clientSideUpload",
+    register: (server) => {
+      server.route({
+        method: "POST",
+        path: "/s3/{id}/{pageKey}/{componentKey}/create-pre-signed-url",
+        handler: async (request: HapiRequest) => {
+          const { uploadService, cacheService } = request.services([]);
+          const state = await cacheService.getState(request);
+          const applicationId = state.metadata?.application_id ?? "";
+          const { id, pageKey, componentKey } = request.params as any;
+          const { filename } = request.payload;
+
+          const form = request.server.app.forms[id];
+          const page = form?.pages.find(
+            (p) =>
+              uploadService.normalisePath(p.path) ===
+              uploadService.normalisePath(pageKey)
+          );
+
+          const metaData = {
+            page: encodeURI(page.title),
+            section: encodeURI(page.section?.title ?? ""),
+            componentName: componentKey,
+          };
+
+          const key = `${applicationId}/${id}/${pageKey}/${componentKey}/${filename}`;
+          const url = await uploadService.getPreSignedUrlS3(key, metaData);
+          return { url };
+        },
+      });
+
+      server.route({
+        method: "GET",
+        path: "/s3/{id}/{pageKey}/{componentKey}/download-file",
```

Comment:
We need to have this query structure so we can securely grab the applicationId from the session cache, this is how we ensure users are only scoped to uploads on their own application:

https://github.com/communitiesuk/digital-form-builder/blob/562aed98d9f4de8b0203349628bb23a566f10a62/runner/src/server/services/cacheService.ts#L144-L155